### PR TITLE
Standalone Lite CI setup

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Foundation for moving Lite to a standalone package.

With this PR, we set up:
- The package
- The CI running tests on the package in separate jobs

Next PRs will move the actual code and replace the placeholder tests cases.

### Does your PR introduce any breaking changes? If yes, please list them.

No.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @carmocca @akihironitta @borda

### CODE REVIEW

File: tests/tests_lite/test_lite.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+from lightning_lite.lite import LightningLite
+
+
+def test_placeholder(tmpdir):
+    assert True
+
+
+@RunIf(min_cuda_gpus=2, standalone=True)
```
Comment (awaelchli) [inline tests/tests_lite/test_lite.py:8]: We can copy the implementation or import it from pl. I leave it up to @carmocca and @Borda to decide what to do here.

Comment (carmocca) [reply tests/tests_lite/test_lite.py:8]: How would you import it? By adding the file to the pythonpath? Tests are usually not importable.

I vote towards copying. It could be imported later from Lite in PL

File: requirements.txt [hunk 2]
```diff
@@ -1,3 +1,4 @@
 # the default package dependencies
 -r ./requirements/app/base.txt
 -r ./requirements/pytorch/base.txt
+-r ./requirements/lite/base.txt
```
Comment (carmocca) [inline requirements.txt:4]: Is this necessary? Do we want the general installation to install Lite by default?

PL would include Lite's dependencies

File: requirements/lite/base.txt [hunk 3]
```diff
@@ -0,0 +1,8 @@
+# NOTE: the upper bound for the package version is only set for CI stability, and it is dropped while installing this package
+#  in case you want to preserve/enforce restrictions on the latest compatible version, add "strict" as an in-line comment
+
+torch>=1.9.*, <1.13.0
+fsspec[http]>=2021.05.0, !=2021.06.0, <2022.6.0
+pyDeprecate>=0.3.1, <=0.3.2
```
Comment (carmocca) [inline requirements/lite/base.txt:6]: Why does Lite need pyDeprecate?

In fact, I think it can be removed from PL too

```python
$ grep -iIrn "pyDeprecate" *
docs/source-pytorch/conf.py:339:    "pyDeprecate": "deprecate",
requirements/pytorch/base.txt:11:pyDeprecate>=0.3.1, <=0.3.2
```

File: tests/tests_lite/run_standalone_tests.sh [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+# THIS FILE ASSUMES IT IS RUN INSIDE THE tests/tests_pytorch DIRECTORY
+
+# Batch size for testing: Determines how many standalone test invocations run in parallel
+# It can be set through the env variable PL_STANDALONE_TESTS_BATCH_SIZE and defaults to 6 if not set
+test_batch_size="${PL_STANDALONE_TESTS_BATCH_SIZE:-6}"
+
+# this environment variable allows special tests to run
+export PL_RUN_STANDALONE_TESTS=1
+# python arguments
+defaults='-m coverage run --source lightning_lite --append -m pytest --no-header'
```
Comment (carmocca) [inline tests/tests_lite/run_standalone_tests.sh:25]: Maybe this could be a symlink where we make the source dir customizable

File: requirements/lite/base.txt (hunk 3 above)
Comment (Borda) [reply requirements/lite/base.txt:6]: sure if we do not use anymore, let's drop it :)

File: tests/tests_lite/test_lite.py (hunk 1 above)
Comment (Borda) [reply tests/tests_lite/test_lite.py:8]: How about outsourcing this class/function to https://github.com/Lightning-AI/devtools
and install it only for testing...

File: requirements.txt (hunk 2 above)
Comment (awaelchli) [reply requirements.txt:4]: I'm just trying to put together a PR, hoping someone will help me get this stuff right.

I already cried for help on Slack because I'm uncertain about these things 😅 . I wish someone else could take these decisions off of me.

Comment (Borda) [reply requirements.txt:4]: happy to help :)

File: tests/tests_lite/test_lite.py (hunk 1 above)
Comment (carmocca) [reply tests/tests_lite/test_lite.py:8]: The class should not be outsourced because each argument is library specific

File: requirements/lite/base.txt (hunk 3 above)
Comment (carmocca) [reply requirements/lite/base.txt:6]: Done in https://github.com/Lightning-AI/lightning/pull/14472

Comment (awaelchli) [reply requirements/lite/base.txt]: Removed here too

File: tests/tests_lite/test_lite.py (hunk 1 above)
Comment (awaelchli) [reply tests/tests_lite/test_lite.py:10]: I copied over a minimal version of RunIf to make sure the basic setup works (min_cuda and standalone). Once the utilities move, we can then copy over more of the implementation with all the different conditions that apply.

File: tests/tests_lite/run_standalone_tests.sh (hunk 4 above)
Comment (awaelchli) [reply tests/tests_lite/run_standalone_tests.sh:25]: Like the batch size through env variable?

Comment (carmocca) [reply tests/tests_lite/run_standalone_tests.sh:25]: That would work

Comment (awaelchli) [reply tests/tests_lite/run_standalone_tests.sh]: Did that. Fingers crossed. [8f37c9d](https://github.com/Lightning-AI/lightning/pull/14451/commits/8f37c9daf28284aa5a68c2743966fa6837e68a72)

**Comment by awaelchli** [comment]
Here are the changes I made to the gpu-tests.yaml file for lite:
[403adad](https://github.com/Lightning-AI/lightning/pull/14451/commits/403adad332e7337a288fbb091afda44cdd0cc480)

@Borda How do I enable it here in the PR?

File: .azure/gpu-tests-lite.yml [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+      # default shm size is 64m. Increase it to avoid:
+      # 'Error while creating shared memory: unhandled system error, NCCL version 2.7.8'
+      options: "--runtime=nvidia -e NVIDIA_VISIBLE_DEVICES=all --shm-size=512m"
+    workspace:
+      clean: all
+
+    steps:
+
+    - bash: |
+        CHANGED_FILES=$(git diff --name-status origin/master -- . | awk  '{print $2}')
```
Comment (carmocca) [inline .azure/gpu-tests-lite.yml:46]: This job does not need this business because we are not going to mark it as required for now. Do you want me to push a commit removing it?

File: .azure/gpu-tests-lite.yml [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    value: '1'
+
+jobs:
+  - job: testing
+    strategy:
+      matrix:
+        'PyTorch - stable':
+          image: "pytorchlightning/pytorch_lightning:base-cuda-py3.9-torch1.12-cuda11.6.1"
+    # how long to run the job before automatically cancelling
+    timeoutInMinutes: "80"
```
Comment (carmocca) [inline .azure/gpu-tests-lite.yml:31]: The timeout should be adjusted, I expect it will take considerably less.

Can be done in a follow-up but I'd add a TODO(lite)

File: .azure/gpu-tests-lite.yml [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+trigger:
+  tags:
+    include:
+      - '*'
+  branches:
+    include:
+      - "master"
+      - "release/*"
+      - "refs/tags/*"
```
Comment (carmocca) [inline .azure/gpu-tests-lite.yml:14]: Run only when these are changed

```suggestion
      - "refs/tags/*"
  paths:
    include:
      - ".azure/gpu-tests-lite.yml"
      - "requirements/lite/**"
      - "src/lightning_lite/**"
      - "tests/tests_lite/**"
```

File: .azure/gpu-tests-lite.yml [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+        PACKAGE_NAME: pytorch
+        FREEZE_REQUIREMENTS: 1
+      displayName: 'Install dependencies'
+      condition: eq(variables['continue'], '1')
+
+    - bash: |
+        set -e
+        python requirements/collect_env_details.py
+        python -c "import torch ; mgpu = torch.cuda.device_count() ; assert mgpu >= 2, f'GPU: {mgpu}'"
+        python requirements/pytorch/check-avail-strategies.py
```
Comment (carmocca) [inline .azure/gpu-tests-lite.yml:89]: Lite does not support Bagua no? It shouldn't install it, which will make this check fail

File: .github/checkgroup.yml [hunk 9]
```diff
@@ -55,8 +55,10 @@ subprojects:
 
   - id: "pytorch_lightning: Azure GPU"
     paths:
-      - ".azure/gpu-tests.yml"
+      - ".azure/gpu-tests-pytorch.yml"
+      - ".azure/gpu-tests-lite.yml"
```
Comment (carmocca) [inline .github/checkgroup.yml:59]: Adding Lite here is not correct as it'd be a separate job.

There needs to be a new `id` for `lightning_lite: Azure GPU`

File: .azure/gpu-tests-lite.yml (hunk 8 above)
Comment (awaelchli) [reply .azure/gpu-tests-lite.yml:89]: Yes, I removed all bagua install references. Are you saying I should delete
python requirements/pytorch/check-avail-strategies.py ?

Comment (carmocca) [reply .azure/gpu-tests-lite.yml:95]: You can delete it or create an equivalent one that is for Lite deps

File: .azure/gpu-tests-lite.yml (hunk 5 above)
Comment (awaelchli) [reply .azure/gpu-tests-lite.yml:52]: Yes, feel free to. I just wanted to avoid running it on app prs because it takes away GPU resources in the queue.

File: .azure/gpu-tests-lite.yml (hunk 6 above)
Comment (awaelchli) [reply .azure/gpu-tests-lite.yml]: Okay, reduced all timeouts in that file

File: .github/checkgroup.yml (hunk 9 above)
Comment (awaelchli) [reply .github/checkgroup.yml]: Okay, I made the new group

File: .github/checkgroup.yml [hunk 10]
```diff
@@ -41,6 +41,15 @@ subprojects:
...
       - "pl-cpu (windows-2022, 3.7, oldest, stable)"
+      - "lite-cpu (macOS-11, 3.10, latest, stable)"
+      - "lite-cpu (macOS-11, 3.7, latest, stable)"
+      - "lite-cpu (macOS-11, 3.7, oldest, stable)"
+      - "lite-cpu (ubuntu-20.04, 3.10, latest, stable)"
+      - "lite-cpu (ubuntu-20.04, 3.7, latest, stable)"
+      - "lite-cpu (ubuntu-20.04, 3.7, oldest, stable)"
+      - "lite-cpu (windows-2022, 3.10, latest, stable)"
+      - "lite-cpu (windows-2022, 3.7, latest, stable)"
+      - "lite-cpu (windows-2022, 3.7, oldest, stable)"
```
Comment (awaelchli) [inline .github/checkgroup.yml:52]: Is this correct?

I changed pl-cpu to lite-cpu in the .github/workflows/ci-lite-test-full.yml file too

**Comment by awaelchli** [comment]
Adapted the cpu jobs for lite here: [5f4a51b](https://github.com/Lightning-AI/lightning/pull/14451/commits/5f4a51b67683a717be1f8aa624c23c72ecb79ca2)

File: .github/workflows/ci-lite-test-full.yml [hunk 11]
```diff
@@ -45,7 +45,7 @@ jobs:
       id: skip
       shell: bash -l {0}
       run: |
-        FILTER='src/pytorch_lightning|requirements/pytorch|tests/tests_pytorch|examples/pl_*'
+        FILTER='src/lightning_lite|requirements/lite|tests/tests_lite'
```
Comment (awaelchli) [inline .github/workflows/ci-lite-test-full.yml:48]: Should this be removed too @carmocca ?

File: .github/workflows/ci-lite-test-full.yml [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+        pip install -e .[test] --upgrade $flag --find-links "https://download.pytorch.org/whl/${url}"
+        pip list
+      shell: bash
+
+    - name: Cache datasets
+      if: ${{ (steps.skip.outputs.continue == '1') }}
+      uses: actions/cache@v3
+      with:
+        path: Datasets
+        key: pl-dataset
```
Comment (carmocca) [inline .github/workflows/ci-lite-test-full.yml]: I don't think Lite will use datasets, so this could be removed?

File: .github/workflows/ci-lite-test-full.yml [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+        url=$(python -c "print('test/cpu/torch_test.html' if '${{matrix.release}}' == 'pre' else 'cpu/torch_stable.html')" 2>&1)
+        pip install -e .[test] --upgrade $flag --find-links "https://download.pytorch.org/whl/${url}"
+        pip list
+      shell: bash
+
+    - name: Cache datasets
+      uses: actions/cache@v3
+      with:
+        path: Datasets
+        key: pl-dataset
```
Comment (carmocca) [inline .github/workflows/ci-lite-test-full.yml:95]: I don't think Lite will use datasets for tests. Maybe remove this?

File: .github/workflows/ci-lite-test-full.yml (hunk 11 above)
Comment (carmocca) [reply .github/workflows/ci-lite-test-full.yml]: Done in https://github.com/Lightning-AI/lightning/pull/14451/commits/fd6df9bb9f93509a788a0a53f64dabde5e6c9b97

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: .github/workflows/ci-lite-test-full.yml (hunk 13 above)
Comment (Borda) [reply .github/workflows/ci-lite-test-full.yml]: agree, keep it lite :)

File: .azure/gpu-tests-lite.yml [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+pr:
+  - "master"
+  - "release/*"
+
+jobs:
+  - job: testing
+    strategy:
+      matrix:
+        'PyTorch - stable':
+          image: "pytorchlightning/pytorch_lightning:base-cuda-py3.9-torch1.12-cuda11.6.1"
```
Comment (Borda) [inline .azure/gpu-tests-lite.yml:31]:
```suggestion
```

File: .azure/gpu-tests-lite.yml [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+      - "requirements/lite/**"
+      - "src/lightning_lite/**"
+      - "tests/tests_lite/**"
+
+pr:
+  - "master"
+  - "release/*"
+
+jobs:
+  - job: testing
```
Comment (Borda) [inline .azure/gpu-tests-lite.yml:27]: any name here - @awaelchli :rabbit:

File: .azure/gpu-tests-lite.yml (hunk 14 above)
Comment (Borda) [reply .azure/gpu-tests-lite.yml:31]: image

File: .azure/gpu-tests-lite.yml [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+      matrix:
+        'PyTorch - stable':
+          image: "pytorchlightning/pytorch_lightning:base-cuda-py3.9-torch1.12-cuda11.6.1"
+    # how long to run the job before automatically cancelling
+    timeoutInMinutes: "20"
+    # how much time to give 'run always even if cancelled tasks' before stopping them
+    cancelTimeoutInMinutes: "2"
+    pool: azure-jirka-spot
+    container:
+      image: $(image)
```
Comment (Borda) [inline .azure/gpu-tests-lite.yml:38]:
```suggestion
      image: "pytorchlightning/pytorch_lightning:base-cuda-py3.9-torch1.12-cuda11.6.1"
```

File: .azure/gpu-tests-lite.yml (hunk 14 above)
Comment (awaelchli) [reply .azure/gpu-tests-lite.yml:31]: Great. Isn't it a problem if we don't use the` pytorchlightning/pytorch_lightning:base-cuda-py3.9-torch1.12-cuda11.6.1` image?

File: requirements.txt (hunk 2 above)
Comment (awaelchli) [reply requirements.txt:4]: Okay thanks for the help. So should I remove the

`-r ./requirements/lite/base.txt`

from this file then?

Comment (carmocca) [reply requirements.txt:4]: I think so :)
